File: streampark-common/src/main/scala/org/apache/streampark/common/util/RedisClient.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.util

import org.apache.streampark.common.constants.Constants
import org.apache.streampark.common.util.Implicits._

import redis.clients.jedis._
import redis.clients.jedis.exceptions.JedisConnectionException

import java.util.concurrent.ConcurrentHashMap

import scala.annotation.meta.getter
import scala.annotation.tailrec
import scala.util.Random

object RedisClient extends Logger {

  @transient
  @getter
  private lazy val pools: ConcurrentHashMap[RedisEndpoint, JedisPool] =
    new ConcurrentHashMap[RedisEndpoint, JedisPool]()

  @transient
  @getter
  private lazy val clusters: ConcurrentHashMap[RedisEndpoint, JedisCluster] =
    new ConcurrentHashMap[RedisEndpoint, JedisCluster]()

  /**
   * Select a random RedisEndpoint to create or get a Redis connection pool
   *
   * @param res
   * @return
   */
  @tailrec
  def connect(endpoints: Array[RedisEndpoint]): Jedis = {
    require(endpoints.length > 0, "[StreamPark] The RedisEndpoint array is empty!!!")
    val index = Random.nextInt().abs % endpoints.length
    try {
      connect(endpoints(index))
    } catch {
      case e: Exception =>
        logger.error(e.getMessage)
        connect(endpoints.drop(index))
    }
  }

  /**
   * Create or get a Redis connection pool
   *
   * @param re
   * @return
   */
  def connect(re: RedisEndpoint): Jedis = {
    val pool = pools.getOrElseUpdate(re, createJedisPool(re))
    var sleepTime: Int = 4
    var conn: Jedis = null
    while (conn == null) {
      try {
        conn = pool.getResource
      } catch {
        case e: JedisConnectionException
            if e.getCause.toString.contains("ERR max number of clients reached") => {
          if (sleepTime < 500) sleepTime *= 2
          Thread.sleep(sleepTime)
        }
        case e: Exception => throw e
      }
    }
    conn
  }

  /**
   * Create a connection pool
   *
   * @param endpoint
   * @return
   */
  def createJedisPool(endpoint: RedisEndpoint): JedisPool = {
    val endpointEn: RedisEndpoint = endpoint.copy(auth = Constants.DEFAULT_DATAMASK_STRING)
    logInfo(s"[StreamPark] RedisClient: createJedisPool with $endpointEn ")
    new JedisPool(
      poolConfig,
      endpoint.host,
      endpoint.port,
      endpoint.timeout,
      endpoint.auth,
      endpoint.db)
  }

  private lazy val poolConfig = {
    val poolConfig: JedisPoolConfig = new JedisPoolConfig()
    // maximum number of connections
    poolConfig.setMaxTotal(1000)
    // maximum number of connections
    poolConfig.setMaxIdle(64)
    // check validity when getting connection, default false
    poolConfig.setTestOnBorrow(true)
    poolConfig.setTestOnReturn(false)
    // check validity at idle, default false
    poolConfig.setTestWhileIdle(false)
    // minimum idle time for eviction connections, default 1800000 milliseconds (30 minutes)
    poolConfig.setMinEvictableIdleTimeMillis(1800000)
    // evicting scan interval (ms), default -1, if negative, do not run the eviction thread
    poolConfig.setTimeBetweenEvictionRunsMillis(30000)
    poolConfig.setNumTestsPerEvictionRun(-1)
    poolConfig
  }

  def connectCluster(res: RedisEndpoint*): JedisCluster = {
    require(res.nonEmpty, "[StreamPark] The RedisEndpoint array is empty!!!")
    val head = res.head
    val cluster = clusters.getOrElseUpdate(
      head, {
        val hostPorts = res.map(r => new HostAndPort(r.host, r.port)).toSet
        new JedisCluster(hostPorts, head.timeout, 1000, 1, head.auth, poolConfig)
      })
    cluster
  }

  def close(): Unit = pools.foreach { case (_, v) => v.close() }
}

File: streampark-common/src/main/java/org/apache/streampark/common/util/RedisClient.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.util;

import org.apache.streampark.common.constants.Constants;

import org.apache.streampark.shaded.org.slf4j.Logger;

import redis.clients.jedis.HostAndPort;
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisCluster;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisPoolConfig;
import redis.clients.jedis.exceptions.JedisConnectionException;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ThreadLocalRandom;

/** Redis connection pool and cluster client factory. */
public final class RedisClient {

    private static final Logger LOG =
        StreamParkLoggerFactory.loggerFactory().getLogger(RedisClient.class.getName());

    private static final ConcurrentHashMap<RedisEndpoint, JedisPool> POOLS = new ConcurrentHashMap<>();

    private static final ConcurrentHashMap<RedisEndpoint, JedisCluster> CLUSTERS = new ConcurrentHashMap<>();

    private static final JedisPoolConfig POOL_CONFIG;

    static {
        JedisPoolConfig poolConfig = new JedisPoolConfig();
        poolConfig.setMaxTotal(1000);
        poolConfig.setMaxIdle(64);
        poolConfig.setTestOnBorrow(true);
        poolConfig.setTestOnReturn(false);
        poolConfig.setTestWhileIdle(false);
        poolConfig.setMinEvictableIdleTimeMillis(1800000);
        poolConfig.setTimeBetweenEvictionRunsMillis(30000);
        poolConfig.setNumTestsPerEvictionRun(-1);
        POOL_CONFIG = poolConfig;
    }

    private RedisClient() {
    }

    public static Jedis connect(RedisEndpoint[] endpoints) {
        if (endpoints.length == 0) {
            throw new IllegalArgumentException("[StreamPark] The RedisEndpoint array is empty!!!");
        }
        int index = ThreadLocalRandom.current().nextInt(endpoints.length);
        try {
            return connect(endpoints[index]);
        } catch (Exception e) {
            LOG.error("[StreamPark] {}", e.getMessage());
            RedisEndpoint[] remaining = new RedisEndpoint[endpoints.length - index - 1];
            System.arraycopy(endpoints, index + 1, remaining, 0, remaining.length);
            if (remaining.length == 0) {
                throw e;
            }
            return connect(remaining);
        }
    }

    public static Jedis connect(RedisEndpoint endpoint) {
        JedisPool pool = POOLS.computeIfAbsent(endpoint, RedisClient::createJedisPool);
        int sleepTime = 4;
        Jedis conn = null;
        while (conn == null) {
            try {
                conn = pool.getResource();
            } catch (JedisConnectionException e) {
                if (e.getCause() != null
                    && e.getCause().toString().contains("ERR max number of clients reached")) {
                    if (sleepTime < 500) {
                        sleepTime *= 2;
                    }
                    try {
                        Thread.sleep(sleepTime);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        throw new RuntimeException(ie);
                    }
                } else {
                    throw e;
                }
            }
        }
        return conn;
    }

    public static JedisPool createJedisPool(RedisEndpoint endpoint) {
        RedisEndpoint endpointEn = endpoint.withAuth(Constants.DEFAULT_DATAMASK_STRING);
        LOG.info("[StreamPark] RedisClient: createJedisPool with {}", endpointEn);
        return new JedisPool(
            POOL_CONFIG,
            endpoint.getHost(),
            endpoint.getPort(),
            endpoint.getTimeout(),
            endpoint.getAuth(),
            endpoint.getDb());
    }

    public static JedisCluster connectCluster(RedisEndpoint... endpoints) {
        if (endpoints.length == 0) {
            throw new IllegalArgumentException("[StreamPark] The RedisEndpoint array is empty!!!");
        }
        RedisEndpoint head = endpoints[0];
        return CLUSTERS.computeIfAbsent(
            head,
            key -> {
                Set<HostAndPort> hostPorts = new HashSet<>();
                for (RedisEndpoint r : endpoints) {
                    hostPorts.add(new HostAndPort(r.getHost(), r.getPort()));
                }
                return new JedisCluster(hostPorts, head.getTimeout(), 1000, 1, head.getAuth(), POOL_CONFIG);
            });
    }

    public static void close() {
        for (Map.Entry<RedisEndpoint, JedisPool> entry : POOLS.entrySet()) {
            entry.getValue().close();
        }
    }
}
